drivers/tee/tee_shm.c

```diff
@@ -299,11 +303,13 @@ struct tee_shm *tee_shm_register(struct tee_context *ctx, unsigned long addr,
 		goto err;
 	}
 
-	rc = teedev->desc->ops->shm_register(ctx, shm, shm->pages,
-					     shm->num_pages, start);
-	if (rc) {
-		ret = ERR_PTR(rc);
-		goto err;
+	if (!(flags & TEE_SHM_OCALL)) {
+		rc = teedev->desc->ops->shm_register(ctx, shm, shm->pages,
```

Comment:
How about letting the callback teedev->desc->ops->shm_register() decide for itself what to do if TEE_SHM_OCALL is set?